File: erettsegi/4_Tarsalgo/tarsalgo.pas
program tarsalgo;
uses crt;

const max = 1000;
Type Adatok = record
	Ora:integer;
	Perc:integer;
	Azon:integer;
	Irany:string;
END;

var fajl:text;
	Lista:array[1..max] of Adatok;
	ossz:integer = 0;
	bentvan:array[1..100] of boolean;
	
Procedure Feladat1();
BEGIN
	Assign(fajl, 'ajto.txt');
	Reset(fajl);
	
	While NOT EOF(fajl) do BEGIN
		ossz := ossz + 1;
		ReadLn(fajl, Lista[ossz].Ora, Lista[ossz].Perc, Lista[ossz].Azon, Lista[ossz].Irany);
	END;
	
	Close(fajl);
END;

Procedure Feladat2();
var i:integer;
	utolso:integer = 0;
BEGIN
	WriteLn('2. feladat:');
	WriteLn('Az első belépő: ', Lista[1].Azon);
	
	for i := 1 to ossz do BEGIN
		if(Lista[i].Irany = ' ki') then BEGIN
			utolso := i;
		END;
	END;
	
	WriteLn('Az utolsó kilépő: ', Lista[utolso].Azon);
END;

Procedure Feladat3();
var i:integer;
	k:integer;
	athaladas:array[1..100] of integer;
BEGIN
	for i := 1 to 100 do BEGIN
		athaladas[i] := 0;
	END;
	
	for i := 1 to 100 do BEGIN
		for k := 1 to ossz do BEGIN
			if(Lista[k].Azon = i) then BEGIN
				athaladas[i] := athaladas[i] + 1;
			END;
		END;
	END;
	
	Assign(fajl, 'athaladas.txt');
	ReWrite(fajl);

	for i := 1 to 100 do BEGIN
		if(athaladas[i] > 0) then BEGIN
			WriteLn(fajl, i, ' ', athaladas[i]);
		END;
	END;
	
	Close(fajl);
END;

Procedure Feladat4();
var i:integer;
	k:integer;
BEGIN
	WriteLn('4. feladat:');
	Write('A végén a társalgóban voltak: '); //1 11 22 24 29 30 35 37
	
	for i := 1 to 100 do BEGIN
		for k := 1 to ossz do BEGIN
			if(Lista[k].Azon = i) then BEGIN
				Case Lista[k].Irany Of
					' be':bentvan[i] := true;
					' ki':bentvan[i] := false;
				END;
			END;
		END;
	END;
	
	for i := 1 to 100 do BEGIN
		if(bentvan[i]) then BEGIN
			Write(i, ' ');
		END;
	END;
	
	WriteLn;
END;

Procedure Feladat5();
var i:integer;
	letszam:integer = 0;
	ora:integer = 0;
	perc:integer = 0;
	max:integer = 0;
BEGIN
	WriteLn('5. feladat:');
	
	for i := 1 to ossz do BEGIN
		Case Lista[i].Irany Of
			' be':letszam := letszam + 1;
			' ki':letszam := letszam - 1;
		END;
		
		if(letszam > max) then BEGIN
			max := letszam;
			ora := Lista[i].Ora;
			perc := Lista[i].Perc;
		END;
	END;
	
	WriteLn('Például ', ora, ':', perc, '-kor voltak a legtöbben a társalgóban.');
END;

Procedure Feladat7(azonosito:integer) Forward;
Procedure Feladat8(azonosito:integer) Forward;

Procedure Feladat6();
var azonosito:integer;
BEGIN
	WriteLn('6. feladat:');
	Write('Adja meg a személy azonosítóját! ');
	ReadLn(azonosito);
	
	Feladat7(azonosito);
END;

Procedure Feladat7(azonosito:integer);
var i:integer;
BEGIN
	WriteLn;
	WriteLn('7. feladat:');
	for i := 1 to ossz do BEGIN
		if(Lista[i].Azon = azonosito) then BEGIN
			if(Lista[i].Irany = ' be') then BEGIN
				Write(Lista[i].Ora, ':', Lista[i].Perc, '-');
			END else BEGIN
				Write(Lista[i].Ora, ':', Lista[i].Perc);
				WriteLn;
			END;
		END;
	END;
	
	WriteLn;
	Feladat8(azonosito);
END;

Procedure Feladat8(azonosito:integer);
var i:integer;
	osszperc:integer = 0;
	perc:integer;
BEGIN
	WriteLn;
	WriteLn('8. feladat:');
	
	for i := 1 to ossz do BEGIN
		if(Lista[i].Azon = azonosito) then BEGIN
			if(Lista[i].Irany = ' be') then BEGIN
				perc := Lista[i].Perc;
			END else BEGIN
				osszperc := osszperc + (Lista[i].Perc - perc);
			END;
			
			WriteLn(osszperc);
		END;
	END;
	
	Write('A(z) ', azonosito, '. személy összesen ', osszperc, ' percet volt bent,');
	Case bentvan[azonosito] Of
		true: Write(' a megfigyelés végén a társalgóban volt.');
		false: Write(' a megfigyelés végén nem volt a társalgóban.');
	END;
END;

BEGIN
	ClrScr;
	
	Feladat1();
	Feladat2();
	WriteLn;
	Feladat3();
	Feladat4();
	WriteLn;
	Feladat5();
	WriteLn;
	Feladat6();
	
	WriteLn;
	WriteLn;
END.

File: erettsegi/4_Tarsalgo/tarsalgo.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Tarsalgo
{
    class Program
    {
        struct Ajto
        {
            public int ora;
            public int perc;
            public int szemely;
            public bool irany;

            public Ajto (int o, int p, int sz, bool i)
            {
                ora = o;
                perc = p;
                szemely = sz;
                irany = i;
            }
        }
        static List<Ajto> ajtok = new List<Ajto>();

        static void Main(string[] args)
        {
            Feladat1();
            Feladat2();
            Feladat3();
            Feladat4();
            Feladat5();
            Feladat6();
        }

        static void Feladat1()
        {
            FileStream stream = File.Open("C:\\Users\\stats\\Desktop\\4_Tarsalgo\\ajto.txt", FileMode.Open);
            StreamReader reader = new StreamReader(stream);

            string sor;
            string[] darabok;
            while(!reader.EndOfStream)
            {
                sor = reader.ReadLine();
                darabok = sor.Split(' ');

                int ora = Convert.ToInt32(darabok[0]);
                int perc = Convert.ToInt32(darabok[1]);
                int szemely = Convert.ToInt32(darabok[2]);
                bool irany = darabok[3] == "be" ? true : false;
                ajtok.Add(new Ajto(ora, perc, szemely, irany));
            }

            reader.Close();
            stream.Close();
        }
        static void Feladat2()
        {
            Console.WriteLine("2. feladat:");
            Console.WriteLine("Az első belépő: " + ajtok[0].szemely);

            int i = ajtok.Count - 1;
            while(i >= 0)
            {
                if(!ajtok[i].irany)
                {
                    Console.WriteLine("Az utolsó kilépő: " + ajtok[i].szemely);
                    break;
                }

                --i;
            }
            Console.WriteLine();
        }
        static void Feladat3()
        {
            int[] alkalom = new int[50];
            foreach(Ajto ajto in ajtok)
            {
                ++alkalom[ajto.szemely];
            }

            FileStream fajl = File.Open("C:\\Users\\stats\\Desktop\\4_Tarsalgo\\athaladas.txt", FileMode.Create);
            StreamWriter writer = new StreamWriter(fajl);

            for(int i = 0; i < 50; i++)
            {
                if(alkalom[i] > 0)
                    writer.WriteLine(i + " " + alkalom[i]);
            }

            writer.Close();
            fajl.Close();
        }
        static void Feladat4()
        {
            Console.WriteLine("4. feladat:");
            bool[] allapot = new bool[50];

            foreach(Ajto ajto in ajtok)
            {
                allapot[ajto.szemely] = ajto.irany;
            }

            for(int i = 0; i < 50; i++)
            {
                if (allapot[i])
                    Console.Write(i + " ");
            }
            Console.WriteLine();
            Console.WriteLine();
        }
        static void Feladat5() 
        {
            int letszam = 0;
            int max = 0;

            int ora = 0;
            int perc = 0;

            foreach(Ajto ajto in ajtok)
            {
                if (ajto.irany) ++letszam;
                else --letszam;

                if(letszam > max)
                {
                    max = letszam;
                    ora = ajto.ora;
                    perc = ajto.perc;
                }
            }

            Console.WriteLine("5. feladat:");
            Console.WriteLine("Például " + ora + ":" + perc + "-kor voltak a legtöbben a társalgóban.");
            Console.WriteLine();
        }
        static void Feladat6()
        {
            Console.WriteLine("6. feladat:");
            Console.WriteLine("Adja meg a személy azonosítóját!");
            int azonosito = Convert.ToInt32(Console.ReadLine());
            Feladat7(azonosito);
            Console.WriteLine();
        }
        static void Feladat7(int szemely)
        {
            foreach(Ajto ajto in ajtok)
            {
                if (ajto.szemely != szemely) continue;
                if (ajto.irany)
                {
                    Console.Write(ajto.ora + ":" + ajto.perc + "-");
                } else
                {
                    Console.Write(ajto.ora + ":" + ajto.perc);
                    Console.WriteLine();
                }
            }
            Console.WriteLine();
            Feladat8(szemely);
        }
        static void Feladat8(int szemely)
        {
            int osszesen = 0;
            int perc = 0;
            bool bent = false;

            foreach(Ajto i in ajtok)
            {
                if (i.szemely != szemely) continue;
                bent = i.irany;

                if (i.irany) perc = i.perc;
                else osszesen += i.perc - perc;
            }

            Console.WriteLine();
            Console.WriteLine("8. feladat:");
            if (bent) Console.WriteLine("A(z) " + szemely + ". személy összesen " + osszesen + " percet volt bent, a megfigyelés végén a társalgóban volt.");
            else Console.WriteLine("A(z) " + szemely + ". személy összesen " + osszesen + " percet volt bent, a megfigyelés végén nem volt a társalgóban.");
        }
    }
}
